src/systemctl/sysv-compat.c

```diff
@@ -65,38 +65,65 @@ int parse_shutdown_time_spec(const char *t, usec_t *ret) {
 
                 *ret = now(CLOCK_REALTIME) + USEC_PER_MINUTE * u;
         } else {
-                char *e = NULL;
-                long hour, minute;
+                char hour_s[6];
+                char min_s[3];
+                char *colon_position;
+                unsigned long hour, minute;
                 struct tm tm = {};
                 time_t s;
                 usec_t n;
+                bool relative = false;
+                int err;
 
-                errno = 0;
-                hour = strtol(t, &e, 10);
-                if (errno > 0 || *e != ':' || hour < 0 || hour > 23)
+                if (t[0] == '+') {
+                        relative = true;
+                        t++;
+                }
+
+                colon_position = strchr(t,':');
+                if (colon_position == NULL)
+                        return -EINVAL;
+
+                strncpy(hour_s, t, colon_position - t);
+                hour_s[5] = 0;
```

Comment:
this results in memory corruption if someone specifies xxxxxx:x as value.

just copy out the number with strndup() or so

Reply:
I've modified it as per your suggestion. strndup is pretty useful here. Of course, that means I also freed the dynamically allocated strings afterward.